Lab03

willshen8 opened the pull request:
Fixes #306 

Review of colleague's PR #322 

#### Changes proposed in this PR:

## - Letter frequency
- Function to implement letter frequency
- Test cases to cover all unicode

willshen24 committed:
Implemented letters lab with test cases

- Solution check for existence first and then increment it in map
- Sort Letter is done by using sort.string()
- Test cases covers all different unicodes and empty inputs

marksomething commented on 03_letters/willshen8/letters.go:
```diff
@@ -0,0 +1,42 @@
+	letterMap := make(map[rune]int)
+
+	for _, value := range s {
+		if letterFrequency, found := letterMap[value]; found {
```
Take advantage of the default behavior of a `map` in go to return the "nil" value (in the case of int, it's 0) to avoid the need for branching.

willshen8 replied:
Thanks for this. I found an even easier way. just do `letterMap[value]++` and go increment it for you and no need to check for nil or 0.

marksomething commented on 03_letters/willshen8/letters.go:
```diff
@@ -0,0 +1,42 @@
+	frequencySlice := []string{}
+
+	for k, v := range m {
+		convertedKeyInStringFormat := string(k)
```
Consider shorter names when in small blocks

willshen8 replied:
Thanks done.

marksomething commented on 03_letters/willshen8/letters.go:
```diff
@@ -0,0 +1,42 @@
+
+	for k, v := range m {
+		convertedKeyInStringFormat := string(k)
+		mapIntToString := strconv.Itoa(v)
```
This is nice, I had used Sprintf for this.

marksomething commented on 03_letters/willshen8/letters_test.go:
```diff
@@ -0,0 +1,48 @@
+		{name: "Normal ASCII inputs", input: "abcabc123123", expected: "1:2 2:2 3:2 a:2 b:2 c:2"},
+		{name: "Special Character inputs", input: "!@#$%^&*!@#$%^&*", expected: "!:2 #:2 $:2 %:2 &:2 *:2 @:2 ^:2"},
+		{name: "Hello World in Chinese", input: "世界您好", expected: "世:1 好:1 您:1 界:1"},
+		{name: "Emojis", input: "\u263A\u263E", expected: "\u263A:1 \u263E:1"},
```
Good to see multibyte unicode testing :)

willshen8 replied:
:)

marksomething commented on 03_letters/willshen8/letters_test.go:
```diff
@@ -0,0 +1,48 @@
+		test.expected = strings.Replace(test.expected, " ", "\n", -1)
+		actual := strings.Join(sortLetters(letters(test.input)), "\n")
+		testResult := reflect.DeepEqual(actual, test.expected)
+		if testResult == false {
```
Have a look at [testify/assert](https://godoc.org/github.com/stretchr/testify/assert)
Its already included in the project module list [go.mod](https://github.com/anz-bank/go-course/blob/master/go.mod)

willshen8 replied:
Thank you, it saved me 5 lines of code!

marksomething reviewed (changes requested):
Nice work! Only a few suggestions.

willshen24 committed:
Changes added after PR review:
-use testify testing library
-removed branching logic for letter mapping
-use better slice declaration for performance tuning

joshcarp commented on 03_letters/willshen8/letters.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func sortLetters(m map[rune]int) []string {
+	frequencySlice := []string{}
```
Go style is to use 
`var frequencySlice []string{}` for initialising slices, 
or 
`frequencySlice := make([]string{}, 0, len(m))` for optimising for performance

joshcarp replied:
https://github.com/golang/go/wiki/CodeReviewComments

willshen8 replied:
Thanks, I'll take up the option 2 for better performance.

joshcarp commented on 03_letters/willshen8/letters_test.go:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+func TestLetterFrequency(t *testing.T) {
```
Great work! 
I guess it's personal preference, bit I prefer table test to be outside testing functions, that way they can be used across multiple testing functions. 
in this case it doesn't really make a difference, although it's nice to keep in mind

willshen8 replied:
Ok, thanks. I'll keep this in mind for future testing.

joshcarp commented on 03_letters/willshen8/letters_test.go:
```diff
@@ -0,0 +1,44 @@
+
+	for _, test := range tests {
+		//replace spaces with newline character to match expected output - easier to read expected output
+		test.expected = strings.Replace(test.expected, " ", "\n", -1)
```
I understand why you did this, although would it make much difference to put `\n` into the test declerations without needing to replace?

willshen8 replied:
I did this for previous labs, and Cameron recommend to do this for readability of test cases. For I have taken this approach instead.

joshcarp replied:
Had a chat to julia and she agrees that this is somewhat unnecessary. 
If you really wanted to format the code you could put it into a slice and use `strings.join`
Anyway, not a vital point, PR is approved

joshcarp commented on 03_letters/willshen8/letters_test.go:
```diff
@@ -0,0 +1,44 @@
+		//replace spaces with newline character to match expected output - easier to read expected output
+		test.expected = strings.Replace(test.expected, " ", "\n", -1)
+		actual := strings.Join(sortLetters(letters(test.input)), "\n")
+		assert.Equal(t, test.expected, actual, "Test Failed! Actual result is not equal to expected")
```
I would say that `"Test Failed! Actual result is not equal to expected"` doesn't really add anything, the test suite will tell us that a test failed

willshen8 replied:
Agreed, the testify library gives you good feedback anyway. So not adding much values.

joshcarp reviewed (changes requested):
Great work, 
Some changes are just personal preference at this point, but really good!

willshen8 committed:
Merge branch 'master' into lab03
